File: 2025/julia/day11.jl
# Day 11: Graph Path Counting with Memoized DFS
# Direct port from day11.el maintaining same data structures (RowMajorMat) and algorithms

include("aoc_common.jl")

function parse_line(line::AbstractString)::Tuple{String, Vector{String}}
    """Parse line into source and destinations."""
    parts = split(line, ':')
    src = String(parts[1])
    dsts = String.(split(strip(parts[2])))
    return (src, dsts)
end

function intern_name!(name::AbstractString, table::Dict{String, Int})
    """Assign ID to name if not already present."""
    if !haskey(table, name)
        table[name] = length(table)
    end
end

function make_node_table(input_str::String)::Dict{String, Int}
    """Create mapping from node names to IDs."""
    table = Dict{String, Int}()
    for line in split(strip(input_str), '\n')
        isempty(line) && continue
        src, dsts = parse_line(line)
        intern_name!(src, table)
        for dst in dsts
            intern_name!(dst, table)
        end
    end
    return table
end

function fill_mat_line!(line::AbstractString, adj::RowMajorMat, table::Dict{String, Int})
    """Fill adjacency matrix for one line."""
    src, dsts = parse_line(line)
    row = table[src]
    for dst in dsts
        col = table[dst]
        adj[row, col] = 1
    end
end

function process_input(input_str::String)::Tuple{Dict{String, Int}, RowMajorMat}
    """Parse input into node table and adjacency matrix."""
    table = make_node_table(input_str)
    n = length(table)
    adj = RowMajorMat(n, n, 0)

    for line in split(strip(input_str), '\n')
        isempty(line) && continue
        fill_mat_line!(line, adj, table)
    end

    return table, adj
end

# Part 1: Simple DFS to count all paths
function dfs(from_node::Int, to_node::Int, adj::RowMajorMat)::Int
    """Count all paths from from_node to to_node."""
    if from_node == to_node
        return 1
    end

    # Sum paths from all children
    total = 0
    for i in 0:adj.cols-1
        if adj[from_node, i] == 1
            total += dfs(i, to_node, adj)
        end
    end

    return total
end

expect("day11_part1", "5")
function day11_part1(input_str::String)::String
    table, adj = process_input(input_str)
    you = table["you"]
    out = table["out"]
    return string(dfs(you, out, adj))
end

# Part 2: DFS with memoization tracking specific nodes visited
function dfs_p2(from_node::Int, to_node::Int, dac_id::Int, fft_id::Int, adj::RowMajorMat)::Int
    """Count paths that visit both dac and fft nodes."""
    memo = Dict{Tuple{Int, Bool, Bool}, Int}()

    function dfs_inner(from_node::Int, to_node::Int, seen_dac::Bool, seen_fft::Bool)::Int
        # Update seen flags
        seen_dac = seen_dac || (from_node == dac_id)
        seen_fft = seen_fft || (from_node == fft_id)

        # Check memo
        key = (from_node, seen_dac, seen_fft)
        if haskey(memo, key)
            return memo[key]
        end

        # Base case
        if from_node == to_node
            result = (seen_dac && seen_fft) ? 1 : 0
        else
            # Recursive case: sum paths from children
            result = 0
            for i in 0:adj.cols-1
                if adj[from_node, i] == 1
                    result += dfs_inner(i, to_node, seen_dac, seen_fft)
                end
            end
        end

        memo[key] = result
        return result
    end

    return dfs_inner(from_node, to_node, false, false)
end

expect("day11_part2", "2")
function day11_part2(input_str::String)::String
    table, adj = process_input(input_str)
    svr = table["svr"]
    out = table["out"]
    dac = table["dac"]
    fft = table["fft"]
    result = dfs_p2(svr, out, dac, fft, adj)
    return string(result)
end

# Main execution
if abspath(PROGRAM_FILE) == @__FILE__
    # Run on real input (no sample file for day11)
    real = read_input(11)
    println("Running Part 1 on real input:")
    println(day11_part1(real))

    println("\nRunning Part 2 on real input:")
    println(day11_part2(real))
end

File: 2025/lua/day11.lua
-- Day 11: Graph Path Counting with Memoized DFS
-- Port from day11.py maintaining same algorithms

local aoc = require("aoc_common")
local Mat = aoc.Mat

local M = {}

-- Parse line into source and destinations
local function parse_line(line)
    local src, dsts_str = line:match("^([^:]+):%s*(.+)$")
    if not src then return nil end

    local dsts = {}
    for dst in dsts_str:gmatch("%S+") do
        table.insert(dsts, dst)
    end

    return src, dsts
end

-- Create mapping from node names to IDs (1-based)
local function make_node_table(input_str)
    local node_map = {}
    local names = {}  -- Keep track of order for counting

    for line in input_str:gmatch("[^\r\n]+") do
        if line ~= "" then
            local src, dsts = parse_line(line)
            if src then
                if not node_map[src] then
                    table.insert(names, src)
                    node_map[src] = #names
                end
                for _, dst in ipairs(dsts) do
                    if not node_map[dst] then
                        table.insert(names, dst)
                        node_map[dst] = #names
                    end
                end
            end
        end
    end

    return node_map, #names
end

-- Fill adjacency matrix for one line
local function fill_mat_line(line, adj, node_map)
    local src, dsts = parse_line(line)
    if not src then return end

    local row = node_map[src]
    for _, dst in ipairs(dsts) do
        local col = node_map[dst]
        adj:set(row, col, 1)
    end
end

-- Parse input into node table and adjacency matrix
local function process_input(input_str)
    local node_map, n = make_node_table(input_str)
    local adj = Mat.new(n, n, 0)

    for line in input_str:gmatch("[^\r\n]+") do
        if line ~= "" then
            fill_mat_line(line, adj, node_map)
        end
    end

    return node_map, adj
end

-- Part 1: Simple DFS to count all paths
local function dfs(from_node, to_node, adj)
    if from_node == to_node then
        return 1
    end

    -- Sum paths from all children
    local total = 0
    for i = 1, adj.cols do
        if adj:ref(from_node, i) == 1 then
            total = total + dfs(i, to_node, adj)
        end
    end

    return total
end

function M.day11_part1(input_str)
    local node_map, adj = process_input(input_str)
    local you = node_map['you']
    local out = node_map['out']
    return tostring(dfs(you, out, adj))
end

-- Part 2: DFS with memoization tracking specific nodes visited
local function dfs_p2(from_node, to_node, dac_id, fft_id, adj)
    local memo = {}

    local function dfs_inner(from_node, to_node, seen_dac, seen_fft)
        -- Update seen flags
        seen_dac = seen_dac or (from_node == dac_id)
        seen_fft = seen_fft or (from_node == fft_id)

        -- Create memo key (encode as single number)
        local key = from_node * 4 + (seen_dac and 2 or 0) + (seen_fft and 1 or 0)
        if memo[key] then
            return memo[key]
        end

        -- Base case
        local result
        if from_node == to_node then
            result = (seen_dac and seen_fft) and 1 or 0
        else
            -- Recursive case: sum paths from children
            result = 0
            for i = 1, adj.cols do
                if adj:ref(from_node, i) == 1 then
                    result = result + dfs_inner(i, to_node, seen_dac, seen_fft)
                end
            end
        end

        memo[key] = result
        return result
    end

    return dfs_inner(from_node, to_node, false, false)
end

function M.day11_part2(input_str)
    local node_map, adj = process_input(input_str)
    local svr = node_map['svr']
    local out = node_map['out']
    local dac = node_map['dac']
    local fft = node_map['fft']
    return tostring(dfs_p2(svr, out, dac, fft, adj))
end

return M
